File: client/Fluentia/Services/CryptoService.cs
using System.Security.Cryptography;
using System.Text;
using Sodium;

namespace Fluentia.Services;

public class CryptoService : IDisposable
{
    private KeyPair _keyPair;
    private byte[]? _peerPublicKey;

    // Receive ratchet (mobile → PC)
    private byte[]? _recvChainKey;
    private int _expectedSeq;
    private bool _ratchetReady;
    private int _highestSeenSeq;
    public const int MaxSeqGap = 1000;

    // Send ratchet (PC → mobile) — backward security
    private byte[]? _sendChainKey;
    private int _sendSeq;
    private bool _sendRatchetReady;

    public CryptoService()
    {
        _keyPair = PublicKeyBox.GenerateKeyPair();
    }

    public string PublicKeyBase64 => Convert.ToBase64String(_keyPair.PublicKey);
    public byte[] ExportPrivateKeyBytes() => (byte[])_keyPair.PrivateKey.Clone();
    public bool IsReady => _peerPublicKey != null;
    public bool RatchetReady => _ratchetReady;
    public bool SendRatchetReady => _sendRatchetReady;

    public void ImportKeyPair(string publicKeyBase64, string privateKeyBase64)
    {
        _keyPair = new KeyPair(
            Convert.FromBase64String(publicKeyBase64),
            Convert.FromBase64String(privateKeyBase64));
        ResetPeerState();
    }

    public void SetPeerPublicKey(string base64Key)
    {
        _peerPublicKey = Convert.FromBase64String(base64Key);
    }

    public void ResetPeerState()
    {
        if (_peerPublicKey != null) { Array.Clear(_peerPublicKey); _peerPublicKey = null; }
        if (_recvChainKey != null) { Array.Clear(_recvChainKey); _recvChainKey = null; }
        _expectedSeq = 0;
        _ratchetReady = false;
        _highestSeenSeq = 0;
        if (_sendChainKey != null) { Array.Clear(_sendChainKey); _sendChainKey = null; }
        _sendSeq = 0;
        _sendRatchetReady = false;
    }

    /// <summary>
    /// Initialize the receive ratchet from the seed sent by mobile.
    /// </summary>
    public void InitRatchet(string seedBase64)
    {
        var seed = Convert.FromBase64String(seedBase64);
        _recvChainKey = HKDF.Expand(HashAlgorithmName.SHA512, seed, 32, Encoding.UTF8.GetBytes("fluentia_chain_v1"));
        _expectedSeq = 0;
        _ratchetReady = true;
        _highestSeenSeq = 0;
    }

    /// <summary>
    /// Initialize the PC's send ratchet for backward security.
    /// Returns the seed that must be sent to mobile.
    /// </summary>
    public string InitSendRatchet()
    {
        var seed = new byte[32];
        RandomNumberGenerator.Fill(seed);
        _sendChainKey = HKDF.Expand(HashAlgorithmName.SHA512, seed, 32, Encoding.UTF8.GetBytes("fluentia_chain_v1"));
        _sendSeq = 0;
        _sendRatchetReady = true;
        return Convert.ToBase64String(seed);
    }

    /// <summary>
    /// Ratcheted encryption (PC → mobile) using SecretBox.
    /// </summary>
    public (string Payload, string Nonce, int Seq) EncryptRatcheted(string plaintext)
    {
        if (_sendChainKey == null)
            throw new InvalidOperationException("Send ratchet not initialized");

        var (msgKey, nextChain) = RatchetStep(_sendChainKey);
        _sendChainKey = nextChain;
        var seq = _sendSeq++;

        var nonce = SecretBox.GenerateNonce();
        var messageBytes = Encoding.UTF8.GetBytes(plaintext);
        var encrypted = SecretBox.Create(messageBytes, nonce, msgKey);

        Array.Clear(msgKey, 0, msgKey.Length);

        return (Convert.ToBase64String(encrypted), Convert.ToBase64String(nonce), seq);
    }

    /// <summary>
    /// Legacy crypto_box encryption.
    /// </summary>
    public (string Payload, string Nonce) Encrypt(string plaintext)
    {
        if (_peerPublicKey == null)
            throw new InvalidOperationException("Peer public key not set");

        var nonce = PublicKeyBox.GenerateNonce();
        var messageBytes = Encoding.UTF8.GetBytes(plaintext);
        var encrypted = PublicKeyBox.Create(messageBytes, nonce, _keyPair.PrivateKey, _peerPublicKey);

        return (Convert.ToBase64String(encrypted), Convert.ToBase64String(nonce));
    }

    /// <summary>
    /// Legacy crypto_box decryption.
    /// </summary>
    public string Decrypt(string payloadBase64, string nonceBase64)
    {
        if (_peerPublicKey == null)
            throw new InvalidOperationException("Peer public key not set");

        var decrypted = PublicKeyBox.Open(
            Convert.FromBase64String(payloadBase64),
            Convert.FromBase64String(nonceBase64),
            _keyPair.PrivateKey,
            _peerPublicKey);

        return Encoding.UTF8.GetString(decrypted);
    }

    /// <summary>
    /// Ratcheted decryption using SecretBox with forward secrecy.
    /// </summary>
    public string DecryptRatcheted(string payloadBase64, string nonceBase64, int seq)
    {
        if (_recvChainKey == null)
            throw new InvalidOperationException("Ratchet not initialized");

        if (seq <= _highestSeenSeq)
            throw new InvalidOperationException($"Replay detected: seq {seq} <= highest seen {_highestSeenSeq}");

        if (seq - _highestSeenSeq > MaxSeqGap)
            throw new InvalidOperationException($"Sequence gap too large: {seq - _highestSeenSeq} > {MaxSeqGap}");

        while (_expectedSeq < seq)
        {
            var (_, next) = RatchetStep(_recvChainKey);
            _recvChainKey = next;
            _expectedSeq++;
        }

        var (msgKey, nextChain) = RatchetStep(_recvChainKey);
        _recvChainKey = nextChain;
        _expectedSeq++;
        _highestSeenSeq = Math.Max(_highestSeenSeq, seq);

        var decrypted = SecretBox.Open(
            Convert.FromBase64String(payloadBase64),
            Convert.FromBase64String(nonceBase64),
            msgKey);

        Array.Clear(msgKey, 0, msgKey.Length);

        return Encoding.UTF8.GetString(decrypted);
    }

    public void Reset()
    {
        var oldPrivate = _keyPair.PrivateKey;
        _keyPair = PublicKeyBox.GenerateKeyPair();
        Array.Clear(oldPrivate);
        ResetPeerState();
    }

    public void Dispose()
    {
        ResetPeerState();
        var privateKey = _keyPair.PrivateKey;
        _keyPair = PublicKeyBox.GenerateKeyPair();
        Array.Clear(privateKey);
        GC.SuppressFinalize(this);
    }

    ~CryptoService()
    {
        Dispose();
    }

    public string CreateBleVerificationCode(string remotePublicKeyBase64)
    {
        var remotePublicKey = Convert.FromBase64String(remotePublicKeyBase64);
        var shared = ScalarMult.Mult(_keyPair.PrivateKey, remotePublicKey);
        var combined = new byte[shared.Length + _keyPair.PublicKey.Length + remotePublicKey.Length];
        Buffer.BlockCopy(shared, 0, combined, 0, shared.Length);
        Buffer.BlockCopy(_keyPair.PublicKey, 0, combined, shared.Length, _keyPair.PublicKey.Length);
        Buffer.BlockCopy(remotePublicKey, 0, combined, shared.Length + _keyPair.PublicKey.Length, remotePublicKey.Length);

        var hash = SHA512.HashData(combined);
        var numeric = ((hash[0] << 16) | (hash[1] << 8) | hash[2]) % 1_000_000;
        return numeric.ToString("D6");
    }

    private static byte[] Kdf(byte[] key, string label)
    {
        var labelBytes = Encoding.UTF8.GetBytes(label);
        var output = new byte[32];
        HKDF.Expand(HashAlgorithmName.SHA512, key, output, labelBytes);
        return output;
    }

    private static (byte[] MessageKey, byte[] NextChainKey) RatchetStep(byte[] chainKey)
    {
        return (Kdf(chainKey, "msg"), Kdf(chainKey, "chain"));
    }
}

File: android/app/src/main/java/com/fluentia/app/crypto/CryptoService.kt
package com.fluentia.app.crypto

import java.util.Base64
import javax.inject.Inject
import javax.inject.Singleton

data class PersistedCryptoSession(
    val publicKey: String,
    val secretKey: String,
    val peerPublicKey: String?,
)

data class EncryptedPayload(
    val payload: String,
    val nonce: String,
    val seq: Int? = null,
)

@Singleton
class CryptoService @Inject constructor(
    private val sodium: SodiumProvider,
) {
    private var keyPair: Pair<ByteArray, ByteArray> // (publicKey, secretKey)
    private var peerPublicKey: ByteArray? = null
    private var ready = false

    private var sendChainKey: ByteArray? = null
    private var sendSeq = 0
    private var ratchetReady = false

    private var recvChainKey: ByteArray? = null
    private var expectedSeq = 0
    private var recvRatchetReady = false

    init {
        keyPair = sodium.cryptoBoxKeypair()
    }

    private fun b64Encode(bytes: ByteArray): String =
        Base64.getEncoder().encodeToString(bytes)

    private fun b64Decode(str: String): ByteArray =
        Base64.getDecoder().decode(str)

    // --- Session persistence ---

    fun importSession(session: PersistedCryptoSession) {
        val pk = b64Decode(session.publicKey)
        val sk = b64Decode(session.secretKey)
        keyPair = pk to sk
        resetPeerState()
        if (session.peerPublicKey != null) {
            peerPublicKey = b64Decode(session.peerPublicKey)
            ready = true
        }
    }

    fun exportSession(): PersistedCryptoSession = PersistedCryptoSession(
        publicKey = b64Encode(keyPair.first),
        secretKey = b64Encode(keyPair.second),
        peerPublicKey = peerPublicKey?.let { b64Encode(it) },
    )

    // --- Public key ---

    fun getPublicKeyBase64(): String = b64Encode(keyPair.first)
    fun getPublicKeyBytes(): ByteArray = keyPair.first.copyOf()
    fun getSecretKeyBytes(): ByteArray = keyPair.second.copyOf()

    fun setPeerPublicKey(base64Key: String) {
        peerPublicKey = b64Decode(base64Key)
        ready = true
    }

    fun hasPeerKey(): Boolean = peerPublicKey != null
    fun isReady(): Boolean = ready
    fun isRatchetReady(): Boolean = ratchetReady
    fun isRecvRatchetReady(): Boolean = recvRatchetReady

    // --- Ratchet init ---

    fun initRatchet(): String {
        val seed = sodium.randomBytes(32)
        sendChainKey = kdf(seed, "fluentia_chain_v1")
        sendSeq = 0
        ratchetReady = true
        return b64Encode(seed)
    }

    fun initRecvRatchet(seedBase64: String) {
        val seed = b64Decode(seedBase64)
        recvChainKey = kdf(seed, "fluentia_chain_v1")
        expectedSeq = 0
        recvRatchetReady = true
    }

    // --- Legacy crypto_box encrypt/decrypt ---

    fun encrypt(plaintext: String): EncryptedPayload {
        val pk = peerPublicKey ?: throw IllegalStateException("Peer public key not set")
        val nonce = sodium.randomBytes(SodiumProvider.BOX_NONCE_BYTES)
        val messageBytes = plaintext.toByteArray(Charsets.UTF_8)
        val cipher = ByteArray(messageBytes.size + SodiumProvider.BOX_MAC_BYTES)
        sodium.cryptoBoxEasy(cipher, messageBytes, nonce, pk, keyPair.second)
        return EncryptedPayload(
            payload = b64Encode(cipher),
            nonce = b64Encode(nonce),
        )
    }

    fun decrypt(payloadBase64: String, nonceBase64: String): String {
        val pk = peerPublicKey ?: throw IllegalStateException("Peer public key not set")
        val cipher = b64Decode(payloadBase64)
        val nonce = b64Decode(nonceBase64)
        val message = ByteArray(cipher.size - SodiumProvider.BOX_MAC_BYTES)
        val ok = sodium.cryptoBoxOpenEasy(message, cipher, nonce, pk, keyPair.second)
        if (!ok) throw RuntimeException("Decryption failed")
        return String(message, Charsets.UTF_8)
    }

    // --- Ratcheted encrypt/decrypt ---

    fun encryptRatcheted(plaintext: String): EncryptedPayload {
        val ck = sendChainKey ?: throw IllegalStateException("Ratchet not initialized")
        val (messageKey, nextChainKey) = ratchetStep(ck)
        sendChainKey = nextChainKey
        val seq = sendSeq++

        val nonce = sodium.randomBytes(SodiumProvider.SECRET_BOX_NONCE_BYTES)
        val messageBytes = plaintext.toByteArray(Charsets.UTF_8)
        val cipher = ByteArray(messageBytes.size + SodiumProvider.SECRET_BOX_MAC_BYTES)
        sodium.cryptoSecretBoxEasy(cipher, messageBytes, nonce, messageKey)
        messageKey.fill(0)

        return EncryptedPayload(
            payload = b64Encode(cipher),
            nonce = b64Encode(nonce),
            seq = seq,
        )
    }

    fun decryptRatcheted(payloadBase64: String, nonceBase64: String, seq: Int): String {
        var ck = recvChainKey ?: throw IllegalStateException("Receive ratchet not initialized")

        while (expectedSeq < seq) {
            val (_, nextChainKey) = ratchetStep(ck)
            ck = nextChainKey
            expectedSeq++
        }
        recvChainKey = ck

        val (messageKey, nextChainKey) = ratchetStep(ck)
        recvChainKey = nextChainKey
        expectedSeq++

        val cipher = b64Decode(payloadBase64)
        val nonce = b64Decode(nonceBase64)
        val message = ByteArray(cipher.size - SodiumProvider.SECRET_BOX_MAC_BYTES)
        val ok = sodium.cryptoSecretBoxOpenEasy(message, cipher, nonce, messageKey)
        messageKey.fill(0)

        if (!ok) throw RuntimeException("Ratcheted decryption failed")
        return String(message, Charsets.UTF_8)
    }

    // --- Reset ---

    fun resetPeerState() {
        peerPublicKey = null
        ready = false
        sendChainKey = null
        sendSeq = 0
        ratchetReady = false
        recvChainKey = null
        expectedSeq = 0
        recvRatchetReady = false
    }

    fun reset() {
        keyPair = sodium.cryptoBoxKeypair()
        resetPeerState()
    }

    // --- KDF: SHA-512(key || label)[0:32] ---

    private fun kdf(key: ByteArray, label: String): ByteArray {
        val labelBytes = label.toByteArray(Charsets.UTF_8)
        val input = key + labelBytes
        val hash = ByteArray(SodiumProvider.SHA512_BYTES)
        sodium.cryptoHashSha512(hash, input)
        return hash.copyOf(32)
    }

    private fun ratchetStep(chainKey: ByteArray): Pair<ByteArray, ByteArray> {
        val messageKey = kdf(chainKey, "msg")
        val nextChainKey = kdf(chainKey, "chain")
        return messageKey to nextChainKey
    }
}
